### ISSUE TITLE
Inconsistent test.py results; P, R, mAP metrics all score (perfect) 1

### ISSUE THREAD

**Issue by cvanlit-supplai** [ultralytics/yolov5#3112]
## ❔What is happening with my test.py results?

After each epoch, the model is tested using the code in test.py. After the last epoch, some plots are saved, including the P-R curve, F1-Curve, confusion matrix, etc.. These plots look realistic to me. On the other hand, when I run 

```
python test.py --img 1280 --conf-thres 0.7 --iou-thres 0.7 --task test --batch-size 2 --device 0 \
--weights runs/MyProject/exp/weights/best.pt
```

(or using `last.pt` and `--task val` for that matter) afterwards, the results look strange. Precision and Recall are 1 for all classes, no matter what `conf-thres` and `iou-thres` I use (even 0.99 and 0.01), and the confusion matrix has 1.00's all along the diagonal axis. Something is obviously going wrong. Any ideas? 

I have included images of the after-last-epoch test and the separate run of `test.py` below.

## Additional context
After-training confusion matrix, as I would expect it to look:
[![confusion-matrix.png](https://i.postimg.cc/R07XkzYr/confusion-matrix.png)](https://postimg.cc/t1g368yD)

When running `test.py` separately:
[![confusion-matrix.png](https://i.postimg.cc/yWCJ4255/confusion-matrix.png)](https://postimg.cc/QKSNcYjk)

Command line output after running `test.py`:
[![test-results.png](https://i.postimg.cc/yY7SV38r/test-results.png)](https://postimg.cc/PNFqyxZm)

**Comment by glenn-jocher** [comment]
@cvanlit-supplai 👋 Hello, thank you for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [test.py](https://github.com/ultralytics/yolov5/blob/master/test.py) in YOLOv5.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, test.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:

## `train.py`

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L188-L192
- **testloader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L199-L202
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## `test.py`

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/test.py#L89-L90
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`

## `detect.py`

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/detect.py#L46-L53
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): False
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`

## YOLOv5 PyTorch Hub

`models.autoShape()` class used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/models/common.py#L225-L250

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): False
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

**Comment by Yukimori-GitHub** [comment]
About the result of 'test.py', I want to know the mean of 'Speed: 61.7/0.4/62.1 ms inference/NMS/total xxxxxx'. What do 61.7, 0.4 and 62.1ms stand for？

**Comment by glenn-jocher** [comment]
@Yukimori-GitHub inference time, nms time, total time in milliseconds per image

**Comment by gaoypeng** [comment]
Hey! I meet the same problem. Have you solved this?

**Comment by glenn-jocher** [comment]
@Hlings you probably have --save-hybrid, which adds the true values to the predictions. If you want accurate mAP, simply run `python val.py`
